src/main/scala/cosas/fns.scala

```diff
@@ -47,4 +47,156 @@ object fns {
 
   trait Fn3[A, B, C] extends AnyFn3 { type In1 = A; type In2 = B; type In3 = C }
 
+
+
+  /* Dependent functions aka dependent products */
+  trait AnyDepFn { type Out }
+  trait AnyDepFn0 extends AnyDepFn
+  trait AnyDepFn1 extends AnyDepFn { type In1 }
+  trait DepFn1[I,O] extends AnyDepFn1 { type In1 = I; type Out = O }
+  trait AnyDepFn2 extends AnyDepFn { type In1; type In2 }
+  trait DepFn2[I1,I2,O] extends AnyDepFn2 { type In1 = I1; type In2 = I2; type Out = O }
+  // TODO moar arities
+  // trait AnyDepFn3 extends AnyDepFn { type In1; type In2 }
+
+  trait AnyDepFn1Composition extends AnyDepFn1 {
+
+    type First <: AnyDepFn1
+    val first: First
+    type Second <: AnyDepFn1 { type In1 >: First#Out }
+    val second: Second
+
+    type In1 = First#In1
+    type Out = Second#Out
+  }
+
+  class Composition[
+    F <: AnyDepFn1,
+    S <: AnyDepFn1 { type In1 >: F#Out }
+  ]
+  (val first: F, val second: S) extends AnyDepFn1Composition {
+
+    type First = F
+    type Second = S
+
+    // implicit def appForMe[
+    //   X1 <: F#In1,
+    //   M <: F#Out,
+    //   O <: S#Out
+    // ](implicit
+    //   appF: App1[F,X1,M],
+    //   appS: App1[S,M,O]
+    // ): App1[this.type,X1,O] = App1 { x1: X1 => appS(appF(x1)) }
+  }
+
+  implicit final def depFn1Ops[DF <: AnyDepFn1, I0 <: DF#In1, O0 <: DF#Out](df: DF): DepFn1Ops[DF,I0,O0] =
+    DepFn1Ops(df)
+  case class DepFn1Ops[DF <: AnyDepFn1, I <: DF#In1, O <: DF#Out](val df: DF) extends AnyVal {
+
+    def at(f: I => O): App1[DF,I,O] =
+      App1(f)
+  }
+
+  implicit final def depFn2Ops[DF <: AnyDepFn2](df: DF): DepFn2Ops[DF] =
+    DepFn2Ops(df)
+  case class DepFn2Ops[DF <: AnyDepFn2](val df: DF) extends AnyVal {
+
+    def at[X1 <: DF#In1, X2 <: DF#In2, Y <: DF#Out](f: (X1,X2) => Y): App2[DF,X1,X2,Y] =
+      App2(f)
+  }
+
+  /* dependent function application machinery. These are to be thought of as the building blocks for terms of a dependent function type. */
+  trait AnyAt {
+
+    type DepFn <: AnyDepFn
+  }
+  trait At[P0 <: AnyDepFn] extends AnyAt {  type DepFn = P0  }
+
+  trait At0[DF0 <: AnyDepFn0] extends AnyAt {
+
+    type DepFn = DF0
+    type Y <: DF0#Out
+    def apply: Y
+  }
+
+  trait At1[DF1 <: AnyDepFn1] extends AnyAt {
+
+    type DepFn = DF1
+    type X1 <: DepFn#In1
+    type Y <: DepFn#Out
+
+    def apply(in: X1): Y
+  }
+
+  trait At2[DF2 <: AnyDepFn2] extends AnyAt {
+
+    type DepFn = DF2
+    type X1 <: DepFn#In1
+    type X2 <: DepFn#In2
+    type Y <: DepFn#Out
+
+    def apply(in1: X1, in2: X2): Y
+  }
+
+  case class App1[DF <: AnyDepFn1, I <: DF#In1, O <: DF#Out](val does: I => O) extends At1[DF] {
```

Comment:
I'm removing most of these traits, they're just legacy from other approaches I tried.